openlane/scripts/klayout/open_design.py

```diff
@@ -13,91 +12,62 @@
-    import pya
-else:
-    import click
-
-    @click.command()
-    @click.option(
-        "-l",
-        "--input-lef",
-        "input_lefs",
-        multiple=True,
-    )
-    @click.option(
-        "-T",
-        "--lyt",
-        required=True,
-        help="KLayout .lyt file",
-    )
-    @click.option(
-        "-P",
-        "--lyp",
-        required=True,
-        help="KLayout .lyp file",
-    )
-    @click.option(
-        "-M",
-        "--lym",
-        required=True,
-        help="KLayout .map (LEF/DEF layer map) file",
-    )
-    @click.argument("input")
-    def cli(**kwargs):
-        args = [
-            "klayout",
-            "-rm",
-            __file__,
-        ]
-        for key, value in kwargs.items():
-            args.append("-rd")
-            if isinstance(value, tuple) or isinstance(value, list):
-                value = ";".join([os.path.abspath(element) for element in value])
-            elif (
-                isinstance(value, str)
-                and os.path.exists(value)
-                and key != "design_name"
-            ):
-                value = os.path.abspath(value)
-
-            args.append(f"{key}={value or 'NULL'}")
+import shlex
+from typing import Tuple
 
-        os.execlp("klayout", *args)
+import pya  # Must be run inside KLayout-- the library version of pya does not include "Application"
+import click
 
-    if __name__ == "__main__":
-        cli()
 
+@click.command()
+@click.option(
+    "-l",
+    "--input-lef",
+    "input_lefs",
+    multiple=True,
+)
+@click.option(
+    "-T",
+    "--lyt",
+    required=True,
+    help="KLayout .lyt file",
+)
+@click.option(
+    "-P",
+    "--lyp",
+    required=True,
+    help="KLayout .lyp file",
+)
+@click.option(
+    "-M",
+    "--lym",
+    required=True,
+    help="KLayout .map (LEF/DEF layer map) file",
+)
+@click.argument("input")
+def open_design(input_lefs: Tuple[str, ...], lyt: str, lyp: str, lym: str, input: str):
+    try:
+        main_window = pya.Application.instance().main_window()
 
-if TYPE_CHECKING:
-    # Dummy data for type-checking
-    input: str = ""
-    input_lefs: str = ""
-    lyp: str = ""
-    lyt: str = ""
-    lym: str = ""
+        tech = pya.Technology()
+        tech.load(lyt)
 
-try:
-    main_window = pya.Application.instance().main_window()
+        layout_options = tech.load_layout_options
+        layout_options.keep_other_cells = True
+        layout_options.lefdef_config.macro_resolution_mode = 1
+        layout_options.lefdef_config.read_lef_with_def = False
+        layout_options.lefdef_config.lef_files = list(input_lefs)
+        layout_options.lefdef_config.map_file = lym
 
-    tech = pya.Technology()
-    tech.load(lyt)
+        cell_view = main_window.load_layout(input, layout_options, 0)
+        layout_view = cell_view.view()
+        layout_view.load_layer_props(lyp)
+    except Exception as e:
+        print(e, file=sys.stderr)
+        exit(1)
```

Comment:
is there no need for `pya.Application.instance().exit(1)` anymore?

Reply:
I think for this one specifically there is, good script